src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,623 @@
+      try {
+        responses = channel.withChannelBlocking(
+            channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+                .batchStat(builder.build()));
+      } catch (InterruptedException e) {
+        return null;
+      } catch (StatusRuntimeException e) {
+        if (e.getStatus().getCode() == Code.CANCELLED && Thread.currentThread().isInterrupted()) {
+          // The current thread is interrupted, meaning that gRPC calls
+          // into the remote output service are cancelled automatically.
+          // Return null, indicating that a fast digest is not
+          // available.
+          //
+          // Throwing an IOException in this case may cause other parts
+          // of Bazel to throw InconsistentFilesystemExceptions, which
+          // leads to spurious build error messages.
+          return null;
+        }
+        throw new IOException(e);
+      }
+      if (responses.getResponsesCount() != 1) {
+        throw new IOException("Remote output service returned an unexpected number of stat responses");
+      }
+      StatResponse response = responses.getResponses(0);
+      if (!response.hasFileStatus()) {
+        // Output service reports that the file does not exist.
+        throw new FileNotFoundException();
+      }
+      FileStatus fileStatus = response.getFileStatus();
+      if (fileStatus.hasFile()) {
+        FileStatus.File regularFileStatus = fileStatus.getFile();
+        return DigestUtil.toBinaryDigest(regularFileStatus.getDigest());
+      }
+      if (fileStatus.hasExternal()) {
+        // Path is a symbolic link that points to a location outside of
+        // the output path. The output service is unable to give a
+        // digest. We must compute it ourselves.
+        return super.getFastDigest(
+            originalOutputPath.getRelative(fileStatus.getExternal().getNextPath()));
+      }
+      throw new IOException(format("Remote output service did not return the status of file %s", path));
+    }
+  }
+
+  @Override
+  @Nullable
+  public FileSystem createActionFileSystem(
+      FileSystem sourceDelegate,
+      PathFragment execRootFragment,
+      String relativeOutputPath,
+      ImmutableList<Root> sourceRoots,
+      ActionInputMap inputArtifactData,
+      Iterable<Artifact> outputArtifacts,
+      boolean trackFailedRemoteReads) {
+    return new GrpcFileSystem(sourceDelegate, execRootFragment.getRelative(relativeOutputPath));
+  }
+
+  @Override
+  public void checkActionFileSystemForLostInputs(FileSystem actionFileSystem, Action action)
+      throws LostInputsActionExecutionException {
+    // There is no need to provide an implementation for this function,
+    // as StartBuild() on the server is supposed to call
+    // FindMissingBlobs() for all remotely tracked files. This ensures
+    // that they cannot disappear during the build.
+  }
+
+  public ListenableFuture<Void> downloadActionResult(ActionResult actionResult) {
```

Comment:
I think this name is either confusing or I am misunderstanding what is going on in here. In my head, I would think that this downloads nothing, but merely it informs the FUSE about the remote files to track, so I would consider naming it something like `injectRemote[Action]Outputs` or something if that's the case.

There is a separate call for prefetching of files for local execution, which I can imagine could internally mean a proactive download so that a later read from FUSE would not need any network traffic.